### PR TITLE
Unpin TF / Keras and fix bugs in autologging integrations for newer versions

### PR DESCRIPTION
**Pull request by dbczumar**
## What changes are proposed in this pull request?

This PR fixes a number of issues in MLflow's TensorFlow & Keras autologging integrations and unpins these libraries for testing purposes in order to catch regressions in the future.

## How is this patch tested?

MLflow autologging tests

## Release Notes

Bug fixes for autologging compatibility with the latest versions of TensorFlow and Keras

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [X] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/keras.py [hunk 1]
```diff
@@ -728,7 +728,7 @@ def _run_and_log_function(self, original, args, kwargs, unlogged_params, callbac
             early_stop_callback = _early_stop_check(tmp_list[callback_arg_index])
             tmp_list[callback_arg_index] += [__MLflowKerasCallback()]
             args = tuple(tmp_list)
-        elif "callbacks" in kwargs:
+        elif kwargs.get("callbacks"):
```
Comment (dbczumar) [inline mlflow/keras.py:731]: `kwargs["callbacks"]` may still be `None`, as this is the default value for the argument. For example, `fit_generator()` calls `fit()` with `callbacks=None` in Keras >= 2.4.0. Many of our other autologging integrations address this case, but this one does not.

File: tests/tensorflow_autolog/test_tensorflow2_autolog.py [hunk 2]
```diff
@@ -221,7 +221,12 @@ def tf_keras_random_data_run_with_callback(
             restore_best_weights=restore_weights,
         )
     else:
-        callback = tf.keras.callbacks.ProgbarLogger(count_mode="samples")
+
+        class CustomCallback(tf.keras.callbacks.Callback):
```
Comment (dbczumar) [inline tests/tensorflow_autolog/test_tensorflow2_autolog.py:225]: As per https://github.com/tensorflow/tensorflow/issues/38618, manually constructing `ProgbarLogger` produces an infinity-related arithmetic error on recent versions of Keras. Accordingly, we swap in a simpler, valid callback here instead.

File: tests/keras_autolog/test_keras_autolog.py [hunk 3]
```diff
@@ -186,18 +186,19 @@ def keras_random_data_run_with_callback(
...
             restore_best_weights=restore_weights,
         )
     else:
-        if fit_variant == "fit_generator":
-            count_mode = "steps"
-        else:
-            count_mode = "samples"
-        callback = keras.callbacks.callbacks.ProgbarLogger(count_mode=count_mode)
+
+        class CustomCallback(keras.callbacks.Callback):
```
Comment (dbczumar) [inline tests/keras_autolog/test_keras_autolog.py:197]: As per https://github.com/tensorflow/tensorflow/issues/38618, manually constructing `ProgbarLogger` produces an infinity-related arithmetic error on recent versions of Keras. Accordingly, we swap in a simpler, valid callback here instead.

File: tests/keras_autolog/test_keras_autolog.py [hunk 4]
```diff
@@ -186,18 +186,19 @@ def keras_random_data_run_with_callback(
     model = create_model()
     if callback == "early":
         # min_delta is set as such to guarantee early stopping
-        callback = keras.callbacks.callbacks.EarlyStopping(
+        callback = keras.callbacks.EarlyStopping(
```
Comment (dbczumar) [inline tests/keras_autolog/test_keras_autolog.py:189]: `keras.callbacks.EarlyStopping` is valid for the versions of Keras we support, whereas `keras.callbacks.callbacks.EarlyStopping` was removed / renamed in 2.4.0.

File: mlflow/utils/autologging_utils.py [hunk 5]
```diff
@@ -70,39 +36,27 @@ def log_fn_args_as_params(fn, args, kwargs, unlogged=[]):  # pylint: disable=W01
...
-        for name in [name for name in defaults.keys() if name in unlogged]:
-            del defaults[name]
-        try_mlflow_log(mlflow.log_params, defaults)
-
-    # Logging the arguments passed by the user
-    args_dict = dict(
-        (param_name, param_val)
-        for param_name, param_val in zip(all_param_names, args)
-        if param_name not in unlogged
+    param_spec = inspect.signature(fn).parameters
```
Comment (dbczumar) [inline mlflow/utils/autologging_utils.py:39]: According to https://docs.python.org/3/library/inspect.html#inspect.getfullargspec, `inspect.signature()` is the recommended utility for method signature inspection in Python 3. `getfullargspec` is kept for compatibility with Python 2 code, which is no longer a concern for us. Further `inspect.getfullargspec` does not return accurate signature information for the latest versions of TensorFlow and Keras.

Accordingly, we drop in `inspect.signature()` in place of the previous `getfullargspec` and clean up the logic a bit.

File: mlflow/tensorflow.py [hunk 6]
```diff
@@ -1058,11 +1064,16 @@ def add_summary(self, *args, **kwargs):
         (EventFileWriterV2, "add_event", add_event),
         (tensorflow.estimator.Estimator, "train", train),
         (tensorflow.keras.Model, "fit", fit),
-        (tensorflow.keras.Model, "fit_generator", fit_generator),
         (tensorflow.estimator.Estimator, "export_saved_model", export_saved_model),
         (tensorflow.estimator.Estimator, "export_savedmodel", export_savedmodel),
         (FileWriter, "add_summary", add_summary),
     ]
+    if LooseVersion(tensorflow.__version__) <= LooseVersion("2.0.0"):
+        # `fit_generator()` is deprecated in TF > 2.0.0 and simply wraps `fit()`.
```
Comment (dbczumar) [inline mlflow/tensorflow.py:1072]: Ideally, in the future, we will make use of a utility similar to scikit-learn's `_SklearnTrainingSession` to guard against nested runs in other ML autologging integrations. For now, this appears to be a quick & perfectly valid fix.

File: mlflow/keras.py [hunk 7]
```diff
@@ -751,9 +751,20 @@ def fit(self, *args, **kwargs):
...
+        `fit_generator()` method calling a patched `fit()` method, we only patch
+        `fit_generator()` in Keras < 2.4.0.
+        """
         original = gorilla.get_original_attribute(keras.Model, "fit_generator")
         unlogged_params = ["self", "generator", "callbacks", "validation_data", "verbose"]
         return _run_and_log_function(self, original, args, kwargs, unlogged_params, 4)
 
     wrap_patch(keras.Model, "fit", fit)
-    wrap_patch(keras.Model, "fit_generator", fit_generator)
+    # `fit_generator()` is deprecated in Keras >= 2.4.0 and simply wraps `fit()`.
```
Comment (dbczumar) [inline mlflow/keras.py:765]: Ideally, in the future, we will make use of a utility similar to scikit-learn's `_SklearnTrainingSession` to guard against nested runs in other ML autologging integrations. For now, this appears to be a quick & perfectly valid fix.

Comment (dbczumar) [inline mlflow/keras.py:765]: Interestingly, Keras >= 2.4.0 pulls in its `Model` class definition from `tensorflow.keras` with a dependency on `tensorflow >= 2.2.0`. This explains why `fit_generator()` is deprecated in Keras 2.4.0, since `tensorflow` deprecated it in TensorFlow 2.1.0.

File: mlflow/fastai.py [hunk 8]
```diff
@@ -405,7 +405,7 @@ def _run_and_log_function(self, original, args, kwargs, unlogged_params, callbac
             callbacks += list(args[callback_arg_index])
             tmp_list[callback_arg_index] += [__MLflowFastaiCallback(self)]
             args = tuple(tmp_list)
-        elif "callbacks" in kwargs:
+        elif kwargs.get("callbacks"):
```
Comment (dbczumar) [inline mlflow/fastai.py:408]: Fixing this logical error since I caught it in https://github.com/mlflow/mlflow/pull/3735/files#r530146856 as well

File: mlflow/tensorflow.py [hunk 9]
```diff
@@ -994,7 +994,7 @@ def fit(self, *args, **kwargs):
                 early_stop_callback = _early_stop_check(tmp_list[5])
                 tmp_list[5], log_dir = _setup_callbacks(tmp_list[5], log_models)
                 args = tuple(tmp_list)
-            elif "callbacks" in kwargs:
+            elif kwargs.get("callbacks"):
```
Comment (dbczumar) [inline mlflow/tensorflow.py:997]: `kwargs["callbacks"]` may still be `None`, as this is the default value for the argument. For example, `fit_generator()` calls `fit()` with `callbacks=None` in TensorFlow >= 2.1.0. Many of our other autologging integrations address this case, but this one does not.

**Review by harupy — APPROVED** [review]
LGTM once the lint errors and `ImportError` for `get_unspecified_default_args` are fixed :)

**Comment by harupy** [comment]
This error:

https://github.com/mlflow/mlflow/pull/3735/checks?check_run_id=1452107744#step:4:1531

seems related to this change:

> Metrics and losses are now reported under the exact name specified by the user (e.g. if you pass metrics=['acc'], your metric will be reported under the string "acc", not "accuracy", and inversely metrics=['accuracy'] will be reported under the string "accuracy".

https://github.com/keras-team/keras/releases/tag/2.3.0

File: mlflow/utils/autologging_utils.py [hunk 10]
```diff
@@ -70,39 +36,27 @@ def log_fn_args_as_params(fn, args, kwargs, unlogged=[]):  # pylint: disable=W01
...
+    # Fetch the parameter names for specified positional arguments from the function
+    # signature & create a mapping from positional argument name to specified value
+    params_to_log = {
+        param_info.name: param_val
+        for param_info, param_val in zip(list(param_spec.values())[len(args) :], args)
+    }
```
Comment (harupy) [inline mlflow/utils/autologging_utils.py:45]:
```suggestion
    params_to_log = {
        param_info.name: param_val
        for param_info, param_val in zip(list(param_spec.values()), args)
    }
```

Should `[len(args) :]` here be removed?

Comment (dbczumar) [reply mlflow/utils/autologging_utils.py:45]: I don't think so, unless I'm missing something. `param_spec.values()` contains all of the methods parameters in order: every positional argument and keyword argument. To find parameter names for only the user-specified `args`, we should slice the first `len(args)` arguments, which are guaranteed to be positional arguments (or positional-or-keyword type arguments). Let me know if this makes sense.

Comment (dbczumar) [reply mlflow/utils/autologging_utils.py:45]: Ah, this should be `:len(args)` though.

Comment (dbczumar) [reply mlflow/utils/autologging_utils.py]: Fixed and added a test case: `test_tf_keras_autolog_names_positional_parameters_correctly`

File: mlflow/utils/autologging_utils.py [hunk 11]
```diff
@@ -70,39 +36,31 @@ def log_fn_args_as_params(fn, args, kwargs, unlogged=[]):  # pylint: disable=W01
...
-
-    # Logging the arguments passed by the user
-    args_dict = dict(
-        (param_name, param_val)
-        for param_name, param_val in zip(all_param_names, args)
-        if param_name not in unlogged
+    param_spec = inspect.signature(fn).parameters
+    # Filter out `self` from the signature under the assumption that it is not specified
+    # as an explicit parameter in the call to `fn`
+    relevant_params = [param for param in param_spec.values() if param.name != "self"]
```
Comment (dbczumar) [inline mlflow/utils/autologging_utils.py:42]: Not a huge fan of this change (i.e. what if `self` is actually the first element of `args`?). Will revisit this tomorrow.

Comment (dbczumar) [reply mlflow/utils/autologging_utils.py:45]: Seems acceptable - callers can filter out `self` via the patched function signature. I've documented the expectation that `self` is not a part of the `args` passed to this function.
